### PR TITLE
feat(module:*): support global config

### PR DESCRIPTION
**Pull request by wendzhue**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Application (the showcase website) / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

### CODE REVIEW

**Comment by wendzhue** [comment]
## Introduction to the resolution

`NzConfigService`. This service would hold the global config and provide methods to get or mutate the config.

`WithConfig` is decorator to help you extract default config for a specific pair of component and property. You should use it with a component name and a default value (optional), and remove the default value assigned value to that property.

File: components/progress/nz-progress.component.ts [hunk 1]
```diff
@@ -31,23 +36,24 @@ export type NzProgressStrokeLinecapType = 'round' | 'square';
...
-  @Input() @InputNumber() nzStrokeWidth: number;
-  @Input() @InputNumber() nzGapDegree: number;
+  @Input() @WithConfig(componentName) @InputNumber() nzStrokeWidth: number;
+  @Input() @WithConfig(componentName) @InputNumber() nzGapDegree: number;
   @Input() nzStatus: NzProgressStatusType;
-  @Input() nzType: NzProgressTypeType = 'line';
-  @Input() nzGapPosition?: NzProgressGapPositionType;
-  @Input() nzStrokeLinecap: NzProgressStrokeLinecapType = 'round';
+  @Input() @WithConfig(componentName, 'line') nzType: NzProgressTypeType;
+  @Input() @WithConfig(componentName, 'top') nzGapPosition?: NzProgressGapPositionType;
```
Comment (hsuanxyz) [inline components/progress/nz-progress.component.ts]: 这里的可选标记`?`可以拿掉吗

File: docs/global-config.en-US.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+1. When you call `NzNotificationService.success`, you passed `{ nzDuration: 6000 }` as the third parameter
+2. You provide `{ nzNotification: { nzDuration: 5000 } }` with `NZ_CONFIG`
+3. You provide  `{ nzDuration: 4000 }` with `NZ_NOTIFICATION_CONFIG`
+4. ng-zorro-antd has a default value of 4500
+
+Eventually, this notification would keep open for 6000 milliseconds.
+
+## Check all Available Globally Configurable Parameters
+
+The interface `Nzconfig` can provide completion suggestions of all components and parameters that are globally configurable. You can also check components' API for more details.
```
Comment (hsuanxyz) [inline docs/global-config.en-US.md]:
```suggestion
The interface `NzConfig` can provide completion suggestions of all components and parameters that are globally configurable. You can also check components' API for more details.
```

File: docs/global-config.zh-CN.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+1. 你在调用 `NzNotificationService.success` 时传递参数 `{ nzDuration: 6000 }`
+2. 你通过 `NZ_CONFIG` 提供了全局默认值 `{ nzNotification: { nzDuration: 5000 } }`
+3. 你通过 `NZ_NOTIFICATION_CONFIG` 提供了全局默认值 `{ nzDuration: 4000 }`
+4. NG-ZORRO 内部默认值为 4500
+
+最终该 Notification 将会显示 6000 毫秒。
+
+## 查看所有可用的全局配置项
+
+`Nzconfig` 接口提供的类型定义信息能够帮助你找到所有支持全局配置项的组件和属性。另外，每个组件的文档都会指出哪些属性可以通过全局配置项的方式指定。
```
Comment (hsuanxyz) [inline docs/global-config.zh-CN.md]:
```suggestion
`NzConfig ` 接口提供的类型定义信息能够帮助你找到所有支持全局配置项的组件和属性。另外，每个组件的文档都会指出哪些属性可以通过全局配置项的方式指定。
```

File: components/button/nz-button.component.ts [hunk 4]
```diff
@@ -72,13 +72,20 @@ export class NzButtonComponent implements AfterContentInit, OnInit, OnDestroy, O
     this.waveConfig,
     this.animationType
   );
-  @Input() @InputBoolean() nzBlock = false;
-  @Input() @InputBoolean() nzGhost = false;
-  @Input() @InputBoolean() nzSearch = false;
-  @Input() @InputBoolean() nzLoading = false;
+
+  @Input() @WithConfig(componentName, false) @InputBoolean() nzBlock: boolean;
```
Comment (hsuanxyz) [inline components/button/nz-button.component.ts]: nzBlock 应该全局配置，使用场景差异很大

File: components/button/doc/index.en-US.md [hunk 5]
```diff
@@ -24,11 +24,11 @@ import { NzButtonModule } from 'ng-zorro-antd/button';
...
 | `[nzGhost]` | make background transparent and invert text and border colors | `boolean` | `false` |
 | `[nzLoading]` | set the loading status of button | `boolean` | `false` |
-| `[nzShape]` | can be set to `circle` `round` or omitted | `'circle' \| 'round'` | - |
-| `[nzSize]` | can be set to `small` `large` or omitted | `'large' \| 'small' \| 'default'` | `'default'` |
-| `[nzType]` | can be set to `primary` `dashed` `danger` or omitted (meaning `default`) | `'primary' \| 'dashed' \| 'danger' \| 'default' \| 'link'` | `'default'` |
-| `[nzBlock]` | option to fit button width to its parent width | `boolean` | `false` |
+| `[nzShape]` | can be set to `circle` `round` or omitted | `'circle'\|'round'` | - |
+| `[nzSize]` | can be set to `small` `large` or omitted | `'large'\|'small'\|'default'` | `'default'` | ✅ |
+| `[nzType]` | can be set to `primary` `dashed` `danger` or omitted (meaning `default`) | `'primary'\|'dashed'\|'danger'\|'default'\|'link'` | `'default'` |
+| `[nzBlock]` | option to fit button width to its parent width | `boolean` | `false` | ✅ |
```
Comment (hsuanxyz) [inline components/button/doc/index.en-US.md]:
```suggestion
| `[nzBlock]` | option to fit button width to its parent width | `boolean` | `false` | - |
```

File: components/button/doc/index.zh-CN.md [hunk 6]
```diff
@@ -28,11 +28,11 @@ import { NzButtonModule } from 'ng-zorro-antd/button';
...
 | `[nzGhost]` | 幽灵属性，使按钮背景透明 | `boolean` | `false` |
 | `[nzLoading]` | 设置按钮载入状态 | `boolean` | `false` |
-| `[nzShape]` | 设置按钮形状，可选值为 `circle` `round` 或者不设 | `'circle' \| 'round'` | - |
-| `[nzSize]` | 设置按钮大小，可选值为 `small` `large` 或者不设 | `'large' \| 'small' \| 'default'` | `'default'` |
-| `[nzType]` | 设置按钮类型，可选值为 `primary` `dashed` `danger` 或者不设 | `'primary' \| 'dashed' \| 'danger' \| 'default' \| 'link'` | `'default'` |
-| `[nzBlock]` | 将按钮宽度调整为其父宽度的选项 | `boolean` | `false` |
+| `[nzShape]` | 设置按钮形状，可选值为 `circle` `round` 或者不设 | `'circle'\|'round'` | - |
+| `[nzSize]` | 设置按钮大小，可选值为 `small` `large` 或者不设 | `'large'\|'small'\|'default'` | `'default'` | ✅ |
+| `[nzType]` | 设置按钮类型，可选值为 `primary` `dashed` `danger` 或者不设 | `'primary'\|'dashed'\|'danger'\|'default'\|'link'` | `'default'` |
+| `[nzBlock]` | 将按钮宽度调整为其父宽度的选项 | `boolean` | `false` | ✅ |
```
Comment (hsuanxyz) [inline components/button/doc/index.zh-CN.md]:
```suggestion
| `[nzBlock]` | 将按钮宽度调整为其父宽度的选项 | `boolean` | `false` | - |
```

File: components/grid/doc/index.en-US.md [hunk 7]
```diff
@@ -91,12 +91,12 @@ import { NzGridModule } from 'ng-zorro-antd/grid';
...
-| `[nzAlign]` | the vertical alignment of the flex layout | `'top' \| 'middle' \| 'bottom'` | `'top'` |
-| `[nzGutter]` | spacing between grids, could be a number or a object like `{ xs: 8, sm: 16, md: 24, lg: 32, xl: 32, xxl: 32 }` | `number \| object` | `0` |
-| `[nzJustify]` | horizontal arrangement of the flex layout | `'start' \| 'end' \| 'center' \| 'space-around' \| 'space-between'` | `'start'` |
-| `[nzType]` | layout mode, optional `flex`, [browser support](http://caniuse.com/#search=flex) | `'flex'` | - |
+| Property | Description | Type | Default | Global Config |
+| -------- | ----------- | ---- | ------- | ------------- |
+| `[nzAlign]` | the vertical alignment of the flex layout | `'top'\|'middle'\|'bottom'` | `'top'` | ✅ |
+| `[nzGutter]` | spacing between grids, could be a number or a object like `{ xs: 8, sm: 16, md: 24, lg: 32, xl: 32, xxl: 32 }` | `number\|object` | `0` | ✅ |
+| `[nzJustify]` | horizontal arrangement of the flex layout | `'start'\|'end'\|'center'\|'space-around'\|'space-between'` | `'start'` | ✅ |
+| `[nzType]` | layout mode, optional `flex`, [browser support](http://caniuse.com/#search=flex) | `'flex'` | - | ✅ |
```
Comment (hsuanxyz) [inline components/grid/doc/index.en-US.md]: 中文文档漏掉了

File: components/progress/doc/index.zh-CN.md [hunk 8]
```diff
@@ -26,17 +26,17 @@ import { NzProgressModule } from 'ng-zorro-antd/progress';
...
-| `[nzShowInfo]` | 是否显示进度数值或状态图标 | `boolean` | `true` |
-| `[nzStatus]` | 状态 | `'success' \| 'exception' \| 'active' \| 'normal'` | - |
-| `[nzStrokeWidth]` `(nzType=line)` | 进度条线的宽度，单位 px | `number` | `8` |
-| `[nzStrokeWidth]` `(nzType=circle)` | 圆形进度条线的宽度，单位是进度条画布宽度的百分比 | `number` | `6` |
-| `[nzType]` | 类型 | `'line' \| 'circle' \| 'dashboard'` | `'line'` |
+| `[nzShowInfo]` | 是否显示进度数值或状态图标 | `boolean` | `true` | ✅ |
+| `[nzStatus]` | 状态 | `'success'\|'exception'\|'active'\|'normal'` | - |
+| `[nzStrokeWidth]` `(nzType=line)` | 进度条线的宽度，单位 px | `number` | `8` | ✅ |
+| `[nzStrokeWidth]` `(nzType=circle)` | 圆形进度条线的宽度，单位是进度条画布宽度的百分比 | `number` | `6` | ✅ |
+| `[nzType]` | 类型 | `'line'\|'circle'\|'dashboard'` | `'line'` | ✅ |
```
Comment (hsuanxyz) [inline components/progress/doc/index.zh-CN.md]: 这里 `nzType`应该不需要全局配置

**Comment by wendzhue** [comment]
This solution is compatible with Ivy. https://github.com/wendzhue/zorro-global-config-ivy-test

File: components/core/config/config.ts [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+  'nzEmpty',
+  'nzIcon',
+  'nzMessage',
+  'nzNotification',
+  'nzProgress',
+  'nzRate'
+] as const;
+
+export interface NzConfig {
+  nzAffix?: NzAffixConfig;
```
Comment (vthinkxie) [inline components/core/config/config.ts]: remove nz prefix here

File: components/button/nz-button.component.ts [hunk 10]
```diff
@@ -143,10 +150,18 @@ export class NzButtonComponent implements AfterContentInit, OnInit, OnDestroy, O
     private contentObserver: ContentObserver,
     private nzUpdateHostClassService: NzUpdateHostClassService,
     private ngZone: NgZone,
+    public nzConfigService: NzConfigService,
     @Optional() @Inject(NZ_WAVE_GLOBAL_CONFIG) private waveConfig: NzWaveConfig,
     @Optional() @Inject(ANIMATION_MODULE_TYPE) private animationType: string
   ) {
     this.renderer.addClass(elementRef.nativeElement, 'ant-btn');
+    this.nzConfigService
+      .getConfigChangeEventForComponent(componentName)
```
Comment (vthinkxie) [inline components/button/nz-button.component.ts]: try this.constructor.name

File: components/tree-select/nz-tree-select.component.ts [hunk 11]
```diff
@@ -99,14 +101,14 @@ export function higherOrderServiceFactory(injector: Injector): NzTreeBaseService
...
 export class NzTreeSelectComponent extends NzTreeBase implements ControlValueAccessor, OnInit, OnDestroy, OnChanges {
-  @Input() @InputBoolean() nzAllowClear = true;
-  @Input() @InputBoolean() nzShowExpand = true;
-  @Input() @InputBoolean() nzShowLine = false;
-  @Input() @InputBoolean() nzDropdownMatchSelectWidth = true;
-  @Input() @InputBoolean() nzCheckable = false;
-  @Input() @InputBoolean() nzHideUnMatched = false;
-  @Input() @InputBoolean() nzShowIcon = false;
-  @Input() @InputBoolean() nzShowSearch = false;
+  @Input() @InputBoolean() @WithConfig(true) nzAllowClear: boolean;
```
Comment (hsuanxyz) [inline components/tree-select/nz-tree-select.component.ts]:
```suggestion
  @Input() @InputBoolean() nzAllowClear: boolean;
```

File: components/tree-select/nz-tree-select.component.ts [hunk 12]
```diff
@@ -139,7 +141,7 @@ export class NzTreeSelectComponent extends NzTreeBase implements ControlValueAcc
   }
 
   @Input() nzDisplayWith: (node: NzTreeNode) => string | undefined = (node: NzTreeNode) => node.title;
-  @Input() nzMaxTagCount: number;
+  @Input() @WithConfig() nzMaxTagCount: number;
```
Comment (hsuanxyz) [inline components/tree-select/nz-tree-select.component.ts]:
```suggestion
  @Input() nzMaxTagCount: number;
```

File: components/tree-select/doc/index.zh-CN.md [hunk 13]
```diff
@@ -23,28 +23,28 @@ import { NzTreeSelectModule } from 'ng-zorro-antd/tree-select';
...
+| `[nzCheckable]` | 节点前添加 Checkbox 复选框 | `boolean` | `false` | ✅ |
+| `[nzShowExpand]` | 节点前添加展开图标 | `boolean` | `true` | ✅ |
+| `[nzShowLine]` | 是否展示连接线 | `boolean` | `false` | ✅ |
 | `[nzAsyncData]` | 是否异步加载(显示加载状态) | `boolean` | `false` |
 | `[nzNodes]` | treeNodes 数据 | `NzTreeNodeOptions[]` | `[]` |
 | `[nzDefaultExpandAll]` | 默认展开所有树节点 | `boolean` | `false` |
 | `[nzExpandedKeys]` | 默认展开指定的树节点 | `string[]` | - |
 | `[nzDisplayWith]` | 如何在输入框显示所选的节点值的方法 | `(node: NzTreeNode) => string` | `(node: NzTreeNode) => node.title` |
-| `[nzMaxTagCount]` | 最多显示多少个 tag | number | - |
+| `[nzMaxTagCount]` | 最多显示多少个 tag | number | - | ✅ |
```
Comment (hsuanxyz) [inline components/tree-select/doc/index.zh-CN.md]:
```suggestion
| `[nzMaxTagCount]` | 最多显示多少个 tag | number | - |  |
```

File: components/tree-select/doc/index.zh-CN.md [hunk 14]
```diff
@@ -23,28 +23,28 @@ import { NzTreeSelectModule } from 'ng-zorro-antd/tree-select';
...
 | `[nzDropdownStyle]` | 下拉菜单的样式 | `{ [key: string]: string; }` | - |
 | `[nzMultiple]` | 支持多选（当设置 nzCheckable 时自动变为true） | `boolean` | `false` |
-| `[nzHideUnMatched]` | 搜索隐藏未匹配的节点 | `boolean` | `false` |
-| `[nzSize]` | 选择框大小 | `'large' \| 'small' \| 'default'` | `'default'` |
-| `[nzCheckable]` | 节点前添加 Checkbox 复选框 | `boolean` | `false` |
-| `[nzShowExpand]` | 节点前添加展开图标 | `boolean` | `true` |
-| `[nzShowLine]` | 是否展示连接线 | `boolean` | `false` |
+| `[nzHideUnMatched]` | 搜索隐藏未匹配的节点 | `boolean` | `false` | ✅ |
+| `[nzSize]` | 选择框大小 | `'large' \| 'small' \| 'default'` | `'default'` | ✅ |
+| `[nzCheckable]` | 节点前添加 Checkbox 复选框 | `boolean` | `false` | ✅ |
```
Comment (hsuanxyz) [inline components/tree-select/doc/index.zh-CN.md]:
```suggestion
| `[nzCheckable]` | 节点前添加 Checkbox 复选框 | `boolean` | `false` |  |
```

File: components/tree-select/doc/index.en-US.md [hunk 15]
```diff
@@ -23,28 +23,28 @@ import { NzTreeSelectModule } from 'ng-zorro-antd/tree-select';
...
+| `[nzCheckable]` | Whether to show checkbox on the treeNodes | `boolean` | `false` | ✅ |
+| `[nzShowExpand]` | Show a Expand Icon before the treeNodes | `boolean` | `true` | ✅ |
+| `[nzShowLine]` | Shows a connecting line | `boolean` | `false` | ✅ |
 | `[nzAsyncData]` | Load data asynchronously (should be used with NzTreeNode.addChildren(...)) | `boolean` | `false` |
 | `[nzNodes]` | Data of the treeNodes | `NzTreeNodeOptions[]` | `[]` |
 | `[nzDefaultExpandAll]` | Whether to expand all treeNodes by default | `boolean` | `false` |
 | `[nzExpandedKeys]` | Default expanded treeNodes | `string[]` | - |
 | `[nzDisplayWith]` | How to display the selected node value in the trigger | `(node: NzTreeNode) => string` | `(node: NzTreeNode) => node.title` |
-| `[nzMaxTagCount]` | Max tag count to show| number | - |
+| `[nzMaxTagCount]` | Max tag count to show| number | - | ✅ |
```
Comment (hsuanxyz) [inline components/tree-select/doc/index.en-US.md]:
```suggestion
| `[nzMaxTagCount]` | Max tag count to show| number | - |  |
```

File: components/tree-select/doc/index.en-US.md [hunk 16]
```diff
@@ -23,28 +23,28 @@ import { NzTreeSelectModule } from 'ng-zorro-antd/tree-select';
...
 | `[nzDropdownStyle]` | To set the style of the dropdown menu | `object` | - |
 | `[nzMultiple]` | Support multiple or not, will be `true` when enable `nzCheckable`. | `boolean` | `false` |
-| `[nzHideUnMatched]` | Hide unmatched nodes while searching | `boolean` | `false` |
-| `[nzSize]` | To set the size of the select input | `'large' \| 'small' \| 'default'` | `'default'` |
-| `[nzCheckable]` | Whether to show checkbox on the treeNodes | `boolean` | `false` |
-| `[nzShowExpand]` | Show a Expand Icon before the treeNodes | `boolean` | `true` |
-| `[nzShowLine]` | Shows a connecting line | `boolean` | `false` |
+| `[nzHideUnMatched]` | Hide unmatched nodes while searching | `boolean` | `false` | ✅ |
+| `[nzSize]` | To set the size of the select input | `'large' \| 'small' \| 'default'` | `'default'` | ✅ |
+| `[nzCheckable]` | Whether to show checkbox on the treeNodes | `boolean` | `false` | ✅ |
```
Comment (hsuanxyz) [inline components/tree-select/doc/index.en-US.md]:
```suggestion
| `[nzCheckable]` | Whether to show checkbox on the treeNodes | `boolean` | `false` |  |
```

File: components/tree-select/doc/index.en-US.md [hunk 17]
```diff
@@ -23,28 +23,28 @@ import { NzTreeSelectModule } from 'ng-zorro-antd/tree-select';
 
 ### nz-tree-select
 
-| Property | Description | Type | Default |
-| -------- | ----------- | ---- | ------- |
-| `[nzAllowClear]` | Whether allow clear | `boolean` | `false` |
+| Property | Description | Type | Default | Global Config |
+| -------- | ----------- | ---- | ------- | ------------- |
+| `[nzAllowClear]` | Whether allow clear | `boolean` | `false` | ✅ |
```
Comment (hsuanxyz) [inline components/tree-select/doc/index.en-US.md]:
```suggestion
| `[nzAllowClear]` | Whether allow clear | `boolean` | `false` |  |
```

File: components/badge/nz-badge.component.ts [hunk 18]
```diff
@@ -63,12 +63,12 @@ export class NzBadgeComponent implements OnInit, AfterViewInit, OnChanges, OnDes
   presetColor: string | null = null;
   count: number;
   @ViewChild('contentElement', { static: false }) contentElement: ElementRef;
-  @Input() @InputBoolean() nzShowZero = false;
+  @Input() @WithConfig(false) @InputBoolean() nzShowZero: boolean;
```
Comment (hsuanxyz) [inline components/badge/nz-badge.component.ts]:
```suggestion
  @Input() @InputBoolean() nzShowZero: boolean;
```

File: components/badge/doc/index.zh-CN.md [hunk 19]
```diff
@@ -33,13 +33,13 @@ import { NzBadgeModule } from 'ng-zorro-antd/badge';
...
+| 参数 | 说明 | 类型 | 默认值 | 全局配置项 |
+| --- | --- | --- | --- | --- |
+| `[nzColor]` | 自定义小圆点的颜色 | string | - | ✅ |
 | `[nzCount]` | 展示的数字，大于 nzOverflowCount 时显示为 `${nzOverflowCount}+`，为 0 时隐藏 | `number \| TemplateRef<void>` | - |
 | `[nzDot]` | 不展示数字，只有一个小红点 | `boolean` | `false` |
 | `[nzShowDot]` | 是否展示小红点 | `boolean` | `true` |
-| `[nzOverflowCount]` | 展示封顶的数字值 | `number` | `99` |
-| `[nzShowZero]` | 当数值为 0 时，是否展示 Badge | `boolean` | `false` |
+| `[nzOverflowCount]` | 展示封顶的数字值 | `number` | `99` | ✅ |
+| `[nzShowZero]` | 当数值为 0 时，是否展示 Badge | `boolean` | `false` | ✅ |
```
Comment (hsuanxyz) [inline components/badge/doc/index.zh-CN.md]:
```suggestion
| `[nzShowZero]` | 当数值为 0 时，是否展示 Badge | `boolean` | `false` | - |
```

File: components/badge/doc/index.en-US.md [hunk 20]
```diff
@@ -32,13 +32,13 @@ import { NzBadgeModule } from 'ng-zorro-antd/badge';
...
+| Property | Description | Type | Default | Global Config |
+| -------- | ----------- | ---- | ------- | ------------- |
+| `[nzColor]` | Customize Badge dot color | string | - | ✅ |
 | `[nzCount]` | Number to show in badge | `number \| TemplateRef<void>` | - |
 | `[nzDot]` | Whether to display a red dot instead of `count` | `boolean` | `false` |
 | `[nzShowDot]` | Whether to display the red dot | `boolean` | `true` |
-| `[nzOverflowCount]` | Max count to show | `number` | `99` |
-| `[nzShowZero]` | Whether to show badge when `count` is zero | `boolean` | `false` |
+| `[nzOverflowCount]` | Max count to show | `number` | `99` | ✅ |
+| `[nzShowZero]` | Whether to show badge when `count` is zero | `boolean` | `false` | ✅ |
```
Comment (hsuanxyz) [inline components/badge/doc/index.en-US.md]:
```suggestion
| `[nzShowZero]` | Whether to show badge when `count` is zero | `boolean` | `false` | - |
```

**Comment by hsuanxyz** [comment]
@cipchk @vthinkxie @wilsoncook @simplejason @wenqi73 各位 Review 一下各自负责的组件，是否有些 input 不适合出现在全局配中，之后再移除会引起 break change。

File: components/button/doc/index.zh-CN.md [hunk 21]
```diff
@@ -28,11 +28,11 @@ import { NzButtonModule } from 'ng-zorro-antd/button';
...
-| 属性 | 说明 | 类型 | 默认值 |
-| --- | --- | --- | --- |
+| 属性 | 说明 | 类型 | 默认值 | 支持全局配置 |
+| --- | --- | --- | --- | --- |
 | `[nzGhost]` | 幽灵属性，使按钮背景透明 | `boolean` | `false` |
 | `[nzLoading]` | 设置按钮载入状态 | `boolean` | `false` |
-| `[nzShape]` | 设置按钮形状，可选值为 `circle` `round` 或者不设 | `'circle' \| 'round'` | - |
-| `[nzSize]` | 设置按钮大小，可选值为 `small` `large` 或者不设 | `'large' \| 'small' \| 'default'` | `'default'` |
-| `[nzType]` | 设置按钮类型，可选值为 `primary` `dashed` `danger` 或者不设 | `'primary' \| 'dashed' \| 'danger' \| 'default' \| 'link'` | `'default'` |
+| `[nzShape]` | 设置按钮形状，可选值为 `circle` `round` 或者不设 | `'circle'\|'round'` | - |
```
Comment (wenqi73) [inline components/button/doc/index.zh-CN.md]:
```suggestion
| `[nzShape]` | 设置按钮形状，可选值为 `circle` `round` 或者不设 | `'circle'\|'round'` | ✅ |
```

File: components/core/config/config.service.ts [hunk 22]
```diff
@@ -0,0 +1,113 @@
+/**
+ * @license
+ * Copyright Alibaba.com All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
+ */
+
+// tslint:disable no-any
```
Comment (wenqi73) [inline components/core/config/config.service.ts]:
```suggestion
```

File: components/core/config/config.service.ts [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+ * config.
+ */
+// tslint:disable-next-line:typedef
+export function WithConfig<T>(innerDefaultValue?: T) {
+  return function ConfigDecorator(target: any, propName: any, originalDescriptor?: TypedPropertyDescriptor<T>): any {
+    const privatePropName = `$$__assignedValue__${propName}`;
+    const componentName = trimComponentName(target.constructor.name) as NzConfigKey;
+
+    if (Object.prototype.hasOwnProperty.call(target, privatePropName)) {
+      console.warn(`The prop "${privatePropName}" is already exist, it will be override by ${name} decorator.`);
```
Comment (wenqi73) [inline components/core/config/config.service.ts]: 没有 `name`

Comment (wendzhue) [reply components/core/config/config.service.ts]: 啥意思？

File: components/rate/doc/index.zh-CN.md [hunk 24]
```diff
@@ -24,10 +24,10 @@ import { NzRateModule } from 'ng-zorro-antd/rate';
...
 ### nz-rate
 
-| 属性 | 说明 | 类型 | 默认值 |
-| --- | --- | --- | --- |
-| `[nzAllowClear]` | 是否允许再次点击后清除 | `boolean` | `true` |
-| `[nzAllowHalf]` | 是否允许半选 | `boolean` | `false` |
+| 属性 | 说明 | 类型 | 默认值 | 支持全局配置 |
+| --- | --- | --- | --- | --- |
+| `[nzAllowClear]` | 是否允许再次点击后清除 | `boolean` | `true` | ✅ |
+| `[nzAllowHalf]` | 是否允许半选 | `boolean` | `false` | ✅ |
```
Comment (hsuanxyz) [inline components/rate/doc/index.zh-CN.md]: 这个属性不适合全局配置

File: components/tabs/doc/index.zh-CN.md [hunk 25]
```diff
@@ -30,19 +30,19 @@ import { NzTabsModule } from 'ng-zorro-antd/tabs';
...
-| `[nzAnimated]` | 是否使用动画切换 Tabs，在 nzTabPosition=top \| bottom 时有效 | `boolean \| {inkBar:boolean, tabPane:boolean}` | `true`, 当 `type="card"` 时为 `false` |
-| `[nzSize]` | 大小，提供 `large` `default` 和 `small` 三种大小 | `'large' \| 'small' \| 'default'` | `'default'` |
+| `[nzAnimated]` | 是否使用动画切换 Tabs，在 nzTabPosition=top \| bottom 时有效 | `boolean \| {inkBar:boolean, tabPane:boolean}` | `true`, 当 `type="card"` 时为 `false` | ✅ |
+| `[nzSize]` | 大小，提供 `large` `default` 和 `small` 三种大小 | `'large' \| 'small' \| 'default'` | `'default'` | ✅ |
 | `[nzTabBarExtraContent]` | tab bar 上额外的元素 | `TemplateRef<void>` | - |
 | `[nzTabBarStyle]` | tab bar 的样式对象 | `object` | - |
-| `[nzTabPosition]` | 页签位置，可选值有 `top` `right` `bottom` `left` | `'top' \| 'right' \| 'bottom' \| 'left'` | `'top'` |
-| `[nzType]` | 页签的基本样式，可选 `line`、`card` 类型 | `'line' \| 'card'` | `'line'` |
-| `[nzTabBarGutter]` | tabs 之间的间隙 | `number` | - |
+| `[nzTabPosition]` | 页签位置，可选值有 `top` `right` `bottom` `left` | `'top' \| 'right' \| 'bottom' \| 'left'` | `'top'` | ✅ |
```
Comment (hsuanxyz) [inline components/tabs/doc/index.zh-CN.md]: 这个属性不适合全局配置

File: components/tree-select/doc/index.zh-CN.md [hunk 26]
```diff
@@ -23,22 +23,22 @@ import { NzTreeSelectModule } from 'ng-zorro-antd/tree-select';
...
 | `[nzDropdownStyle]` | 下拉菜单的样式 | `{ [key: string]: string; }` | - |
 | `[nzMultiple]` | 支持多选（当设置 nzCheckable 时自动变为true） | `boolean` | `false` |
-| `[nzHideUnMatched]` | 搜索隐藏未匹配的节点 | `boolean` | `false` |
-| `[nzSize]` | 选择框大小 | `'large' \| 'small' \| 'default'` | `'default'` |
+| `[nzHideUnMatched]` | 搜索隐藏未匹配的节点 | `boolean` | `false` | ✅ |
+| `[nzSize]` | 选择框大小 | `'large' \| 'small' \| 'default'` | `'default'` | ✅ |
 | `[nzCheckable]` | 节点前添加 Checkbox 复选框 | `boolean` | `false` |
-| `[nzShowExpand]` | 节点前添加展开图标 | `boolean` | `true` |
-| `[nzShowLine]` | 是否展示连接线 | `boolean` | `false` |
+| `[nzShowExpand]` | 节点前添加展开图标 | `boolean` | `true` | ✅ |
```
Comment (hsuanxyz) [inline components/tree-select/doc/index.zh-CN.md]: 这个属性不适合全局配置

File: components/tree-select/doc/index.zh-CN.md [hunk 27]
```diff
@@ -23,22 +23,22 @@ import { NzTreeSelectModule } from 'ng-zorro-antd/tree-select';
...
+| `[nzShowLine]` | 是否展示连接线 | `boolean` | `false` | ✅ |
```
Comment (hsuanxyz) [inline components/tree-select/doc/index.zh-CN.md]: 这个属性不适合全局配置

File: components/tree/doc/index.zh-CN.md [hunk 28]
```diff
@@ -21,20 +21,20 @@ import { NzTreeModule } from 'ng-zorro-antd/tree';
...
-| --- | --- | --- | --- |
+| 参数 | 说明 | 类型 | 默认值 | 全局配置 |
+| --- | --- | --- | --- | --- |
 | `[nzData]` | 元数据 | `NzTreeNodeOptions[] \| NzTreeNode[]` | `[]` |
-| `[nzBlockNode]` | 是否节点占据一行 | `boolean` | `false` |
+| `[nzBlockNode]` | 是否节点占据一行 | `boolean` | `false` | ✅ |
 | `[nzCheckable]` | 节点前添加 Checkbox 复选框 | `boolean` | `false` |
-| `[nzShowExpand]` | 节点前添加展开图标 | `boolean` | `true` |
-| `[nzShowLine]` | 是否展示连接线 | `boolean` | `false` |
+| `[nzShowExpand]` | 节点前添加展开图标 | `boolean` | `true` | ✅ |
```
Comment (hsuanxyz) [inline components/tree/doc/index.zh-CN.md]: 这个属性不适合全局配置

File: components/tree/doc/index.zh-CN.md [hunk 29]
```diff
@@ -21,20 +21,20 @@ import { NzTreeModule } from 'ng-zorro-antd/tree';
...
+| `[nzShowLine]` | 是否展示连接线 | `boolean` | `false` | ✅ |
```
Comment (hsuanxyz) [inline components/tree/doc/index.zh-CN.md]: 这个属性不适合全局配置

File: components/button/doc/index.zh-CN.md [hunk 30]
```diff
@@ -28,11 +28,11 @@ import { NzButtonModule } from 'ng-zorro-antd/button';
...
-| 属性 | 说明 | 类型 | 默认值 |
-| --- | --- | --- | --- |
+| 属性 | 说明 | 类型 | 默认值 | 支持全局配置 |
+| --- | --- | --- | --- | --- |
 | `[nzGhost]` | 幽灵属性，使按钮背景透明 | `boolean` | `false` |
 | `[nzLoading]` | 设置按钮载入状态 | `boolean` | `false` |
-| `[nzShape]` | 设置按钮形状，可选值为 `circle` `round` 或者不设 | `'circle' \| 'round'` | - |
-| `[nzSize]` | 设置按钮大小，可选值为 `small` `large` 或者不设 | `'large' \| 'small' \| 'default'` | `'default'` |
-| `[nzType]` | 设置按钮类型，可选值为 `primary` `dashed` `danger` 或者不设 | `'primary' \| 'dashed' \| 'danger' \| 'default' \| 'link'` | `'default'` |
+| `[nzShape]` | 设置按钮形状，可选值为 `circle` `round` 或者不设 | `'circle'\|'round'` | - | ✅ |
```
Comment (hsuanxyz) [inline components/button/doc/index.zh-CN.md]: 这个属性不适合全局配置

**Pull request merged**